File: Question 0039 - Combination Sum/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution
{
    private void solve(int[] candidates, int target, List<List<Integer>> ans, List<Integer> combi, int index)
    {
        //base case
        if(target == 0)
        {
            ans.add(new ArrayList<>(combi));
            return;
        }

        if(index < 0 || target<0) return;

        //include
        combi.add(candidates[index]);
        solve(candidates, target-candidates[index], ans, combi, index);
        combi.removeLast();

        //exclude
        solve(candidates, target, ans, combi, index-1);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target)
    {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> combi = new ArrayList<>();
        solve(candidates,target,ans,combi, candidates.length-1);
        return ans;
    }
}

File: Question 0039 - Combination Sum/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

void solve(vector<vector<int>>& ans, vector<int>& combi, vector<int>& candidates, int target, int idx)
{
    if(target == 0)
    {
        ans.push_back(combi);
        return;
    }

    if(idx < 0 || target < 0) return;
    //take curr
    combi.push_back(candidates[idx]);
    solve(ans, combi, candidates, target-candidates[idx], idx);
    combi.pop_back();

    //exclude
    solve(ans, combi, candidates, target, idx-1);
}

class Solution 
{
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) 
    {
        vector<vector<int>> ans;
        vector<int> combi;
        solve(ans, combi, candidates, target, candidates.size()-1);
        return ans;
    }
};